macrobot/macrobot/handler.go

```diff
@@ -0,0 +1,176 @@
+		h.ChatEcho(msg.ConvID, userErr)
+		return nil
+	}
+
+	switch {
+	case strings.HasPrefix(cmd, "!macro run"):
+		return h.handleRun(msg, tokens[2:])
+	case strings.HasPrefix(cmd, "!macro create"):
+		return h.handleCreate(msg, tokens[2:])
+	case strings.HasPrefix(cmd, "!macro list"):
+		return h.handleList(msg)
+	case strings.HasPrefix(cmd, "!macro remove"):
+		return h.handleRemove(msg, tokens[2:])
+	default:
+		h.ChatEcho(msg.ConvID, "Unknown command %q", cmd)
+		return nil
+	}
+}
+
+func (h *Handler) handleRun(msg chat1.MsgSummary, args []string) error {
+	if len(args) != 1 {
+		h.ChatEcho(msg.ConvID, "Invalid number of arguments. Expected one: <name>")
+		return nil
+	}
+
+	macroName := args[0]
+	macroMessage, err := h.db.Get(msg.Channel, macroName)
+	switch err {
+	case nil:
+	case sql.ErrNoRows:
+		h.ChatEcho(msg.ConvID, "Macro '%s' is not defined for this %s", macroName, getChannelType(msg.Channel))
+		return nil
+	default:
+		return err
+	}
+	sanitizedMacroMessage := sanitizeMessage(macroMessage)
+
+	h.ChatEcho(msg.ConvID, sanitizedMacroMessage)
+	return nil
+}
+
+func (h *Handler) handleCreate(msg chat1.MsgSummary, args []string) error {
+	if len(args) != 2 {
+		h.ChatEcho(msg.ConvID, "Invalid number of arguments. Expected two: <name> <message>")
+		return nil
+	}
+
+	macroName := args[0]
+	macroMessage := args[1]
+	err := h.db.Create(msg.Channel, macroName, macroMessage)
+	if err != nil {
+		return err
+	}
+	h.ChatEcho(msg.ConvID, "Macro '%s' created", macroName)
+	return nil
+}
+
+func (h *Handler) handleList(msg chat1.MsgSummary) error {
+	macroList, err := h.db.List(msg.Channel)
+	if err != nil {
+		return err
+	}
+
+	if len(macroList) == 0 {
+		h.ChatEcho(msg.ConvID, "There are no macros defined for this %s", getChannelType(msg.Channel))
+		return nil
+	}
+
+	data := []interface{}{getChannelType(msg.Channel)}
+	for _, macro := range macroList {
+		data = append(data, macro.Name)
+		data = append(data, macro.Message)
+	}
+	calendarListMessage := "Here are the macros available for this %s:" + strings.Repeat("\n• %s: `%s`", len(macroList))
+	h.ChatEcho(msg.ConvID, calendarListMessage, data...)
+	return nil
+}
+
+func (h *Handler) handleRemove(msg chat1.MsgSummary, args []string) error {
+	if len(args) != 1 {
+		h.ChatEcho(msg.ConvID, "Invalid number of arguments. Expected one: <name>")
+		return nil
+	}
+
+	macroName := args[0]
+	err := h.db.Remove(msg.Channel, macroName)
+	if err != nil {
+		return err
+	}
+
+	h.ChatEcho(msg.ConvID, "Macro '%s' removed", macroName)
+	return nil
+}
+
+func getChannelType(channel chat1.ChatChannel) string {
+	if channel.MembersType == "team" {
+		return "team"
+	} else {
+		return "conversation"
+	}
+}
+
+func sanitizeMessage(message string) string {
+	if strings.HasPrefix(message, "/") && !isWhiteListed(message) {
+		// escape beginning slash
+		message = "\\" + message
+	}
+	// prevent stellar payments by escaping all '+'s
+	message = strings.ReplaceAll(message, "+", "\\+")
```

Comment:
this isn't really necessary since these are restricted at the API level through the `ConfirmLumenSend` parameter.

Reply:
True, although the bot won't succeed at sending the message at all. So if I registered:
```
!macro create lottery '/flip +50xlm@marceloneil, +50xlm@joshblum'
```
Running `!macro run lottery` wouldn't output anything. I guess it could be argued that in the initial definition of the function you should comment out the '+'s though, what do you think?

Reply:
I didn't realize the message wouldn't send, I thought the in-chat-send just wouldn't be parsed. If the message fails to send let's leave the escaping in

Reply:
Maybe that's a bug! I can send you the error message if you're interested.